# Case Study 1: IPL Teams Analysis

## Context
This case study analyzes the IPL 2025 player auction dataset to understand team spending patterns, player valuation, and pricing differences across roles and player categories. Each row represents a player signed by a team, including price (in crore INR), role, acquisition method (Retained, Auction, RTM), and player type (Indian/Overseas, capped/uncapped).

## Schema
**Schema:** `case_study_1`
**Table:** `ipl_players`

**Columns:**
- `player` (TEXT): Player name
- `price_in_cr` (DECIMAL): Auction/retention price in crore INR
- `type` (TEXT): Player category, e.g., Indian/Overseas and capped/uncapped
- `acquisition` (TEXT): Acquisition mode (`Retained`, `Auction`, `RTM`)
- `role` (TEXT): Playing role, e.g., Batter, Bowler, All-rounder, Batter/Wicketkeeper
- `team` (TEXT): IPL franchise name

## Questions
1. Find the total spending on players for each team.
2. Find the top 3 highest-paid all-rounders across all teams.
3. Find the highest-priced player in each team.
4. Rank players by their price within each team and list the top 2 for every team.
5. Find the most expensive player from each team, along with the second-most expensive player's name and price.
6. Calculate the percentage contribution of each player's price to the team's total spending.
7. Classify players as `High`, `Medium`, or `Low` priced based on defined price bands.
8. Find the average price of Indian players and compare it with overseas players using a subquery.
9. Identify players who earn more than the average price of their team.
10. For each role, find the most expensive player and their price using a correlated subquery.

## Setup: connect to Postgres

In [2]:
import os
from dotenv import load_dotenv
from sqlalchemy import create_engine

load_dotenv(dotenv_path='../.env')

DB_USER = os.getenv('DB_USER', 'postgres')
DB_PASSWORD = os.getenv('DB_PASSWORD', '')
DB_HOST = os.getenv('DB_HOST', 'localhost')
DB_PORT = os.getenv('DB_PORT', '5432')
DB_NAME = os.getenv('DB_NAME', 'sql_case_studies')

CONN_STR = f'postgresql+psycopg2://{DB_USER}:{DB_PASSWORD}@{DB_HOST}:{DB_PORT}/{DB_NAME}'
engine = create_engine(CONN_STR)
print('Connecting to:', f'postgresql://{DB_USER}:***@{DB_HOST}:{DB_PORT}/{DB_NAME}')

Connecting to: postgresql://postgres:***@localhost:5432/sql_case_studies


In [3]:
%load_ext sql
%config SqlMagic.autopandas = True
%config SqlMagic.feedback = False
%config SqlMagic.displaycon = False
%sql engine

The 'toml' package isn't installed. To load settings from pyproject.toml or ~/.jupysql/config, install with: pip install toml

In [4]:
%%sql
SELECT version();

,version
0,"PostgreSQL 17.6 on x86_64-windows, compiled by..."


In [5]:
%%sql
SET search_path = case_study_1;
SELECT COUNT(*) AS total_players, COUNT(DISTINCT team) AS teams
FROM ipl_players;

,total_players,teams
0,228,10


In [6]:
%%sql
SELECT column_name, data_type
FROM information_schema.columns
WHERE table_schema = 'case_study_1'
  AND table_name = 'ipl_players'
ORDER BY ordinal_position;

,column_name,data_type
0,player,text
1,price_in_cr,numeric
2,type,text
3,acquisition,text
4,role,text
5,team,text


---
## Q1. Find the total spending on players for each team:

**Thought process:** To find the total spending on players by team we will calculate the sum of price_in_cr for each team (so GROUP BY team) and we want to show the results for highest spending teams to lowest spending teams - ORDER BY total spending DESC.

In [7]:
%%sql
SELECT
    team
    , SUM(price_in_cr) AS total_spending
FROM ipl_players
GROUP BY 1
ORDER BY 2 DESC;

,team,total_spending
0,Lucknow Super Giants,119.90
1,Gujarat Titans,119.85
2,Mumbai Indians,119.80
3,Sunrisers Hyderabad,119.80
4,Punjab Kings,119.65
5,Royal Challengers Bengaluru,119.25
6,Chennai Super Kings,119.20
7,Rajasthan Royals,118.95
8,Delhi Capitals,116.55
9,Kolkata Knight Riders,107.95


**Answer:** 

We can see the top 7 highest spending teams all have similar spending amount of 119 crore.  Whereas the lowest spending team - Kolkata Knight Riders spent a total of 107.95 crore. That is only 12 crore difference between the top and the bottom spenders. However in context, 1 core = $105,250 USD, so even it appears to have small range in Core terms but that is a $1,263,000 USD difference there (and that is huge!). 

Moreover, the distribution of the total spending across team would be left skewed with most data points concentrated in the right side and Kolkata Knight Riders on the left side.

---
# Q2. Find the top 3 highest-paid all rounders across all teams:

Thought process: We want to filter to keep the records for all rounder role. To add more context, we want to show results for their name, type, team and price_in_cr and ORDER BY price_in_cr in DESC order and lastly LIMIT by top 3 rows only.

In [9]:
%%sql
SELECT
    player
    , type
    , team
    , price_in_cr
FROM ipl_players
WHERE role = 'All-rounder'
ORDER BY 4 DESC
LIMIT 3
;

,player,type,team,price_in_cr
0,Ravindra Jadeja,Indian (capped),Chennai Super Kings,18.00
1,Hardik Pandya,Indian (capped),Mumbai Indians,16.35
2,Abhishek Sharma,Indian (capped),Sunrisers Hyderabad,14.00


**Answer:** Our top 3 highest all rounder players are Ravindra, Hardik and Abhishek respectively. They are all type Indian (capped), playing for different teams and have about 2 Crore variance in price. Not to our surprise the highest paid all rounders are in the the top (7) highest spending team discovered earlier.

---
# Q3: Find the highest-priced player in each team:

**Thought process:** We can find the highest-priced player in each team using window function - the first_value() function PARTITION BY each team ORDER BY the highest price to the lowest within each window. 

Note that with this approach we will only return the first_value see (if multiple player has the same value it will still return the first row of the table. So if there are 10 teams the query will return 10 records). If we want to return all rows with the same value we can use DENSE_RANK() instead.

In [34]:
%%sql
SELECT DISTINCT
    team
    , first_value(player) OVER(PARTITION BY team ORDER BY price_in_cr DESC) AS highest_priced_player
    , first_value(price_in_cr) OVER(PARTITION BY team ORDER BY price_in_cr DESC) AS value
FROM ipl_players
ORDER BY 3 DESC;

,team,highest_priced_player,value
0,Lucknow Super Giants,Rishabh Pant,27.00
1,Punjab Kings,Shreyas Iyer,26.75
2,Kolkata Knight Riders,Venkatesh Iyer,23.75
3,Sunrisers Hyderabad,Heinrich Klaasen (South Africa),23.00
4,Royal Challengers Bengaluru,Virat Kohli,21.00
5,Chennai Super Kings,Ruturaj Gaikwad,18.00
6,Gujarat Titans,Rashid Khan (Afghanistan),18.00
7,Mumbai Indians,Jasprit Bumrah,18.00
8,Rajasthan Royals,Yashasvi Jaiswal,18.00
9,Delhi Capitals,Axar Patel,16.50


---
# Q4: Rank players by their price within each team and list the top 2 for every team

Thought process: Okay so this question is similar to the previous one, but this time we will return 2 highest priced players instead of one. Assume the question require us to return strictly 2 players per team (ignoring the fact that there could be more than 2 players with the same highest price) - we will use ROW_NUMBER() in this case.

So first we will find out the row_number for each team based on their price_in_cr. Then we will filter to keep only rank 1 and 2 rows for each team.

In [40]:
%%sql
SELECT
    team
    , player
    , price_in_cr
FROM
    (SELECT
        team
        , player
        , price_in_cr
        , ROW_NUMBER() OVER(PARTITION BY team ORDER BY price_in_cr DESC) AS price_rank
    FROM ipl_players)
WHERE price_rank <= 2;

,team,player,price_in_cr
0,Chennai Super Kings,Ruturaj Gaikwad,18.00
1,Chennai Super Kings,Ravindra Jadeja,18.00
2,Delhi Capitals,Axar Patel,16.50
3,Delhi Capitals,KL Rahul,14.00
4,Gujarat Titans,Rashid Khan (Afghanistan),18.00
5,Gujarat Titans,Shubman Gill,16.50
6,Kolkata Knight Riders,Venkatesh Iyer,23.75
7,Kolkata Knight Riders,Rinku Singh,13.00
8,Lucknow Super Giants,Rishabh Pant,27.00
9,Lucknow Super Giants,Nicholas Pooran (West Indies),21.00


---
# Q5: Find the most expensive player from each team, along with the second-most expensive player's name and price

Thought process: So this question is somewhat identical with Q4 apart from the formatting of the result. So we want for each team show the name of the most expensive player, price_of_most_expensive, second_most_expensive_player's name and price. 

In this case we can use the same technique as above, find the rankings, and then create each column using CASE WHEN logic, and the MAX() and GROUP BY team to return single row for each team.

In [47]:
%%sql
with price_ranking AS
    (SELECT
            team
            , player
            , price_in_cr
            , ROW_NUMBER() OVER(PARTITION BY team ORDER BY price_in_cr DESC) AS price_rank
        FROM ipl_players)

SELECT
    team
    , MAX(CASE WHEN price_rank = 1 THEN player END) AS most_expensive_player
    , MAX(CASE WHEN price_rank = 1 THEN price_in_cr END) AS most_expensive_price
    , MAX(CASE WHEN price_rank = 2 THEN player END) AS second_most_expensive_player
    , MAX(CASE WHEN price_rank = 2 THEN price_in_cr END) AS second_most_expensive_price
FROM price_ranking
GROUP BY team

,team,most_expensive_player,most_expensive_price,second_most_expensive_player,second_most_expensive_price
0,Chennai Super Kings,Ruturaj Gaikwad,18.00,Ravindra Jadeja,18.00
1,Delhi Capitals,Axar Patel,16.50,KL Rahul,14.00
2,Gujarat Titans,Rashid Khan (Afghanistan),18.00,Shubman Gill,16.50
3,Kolkata Knight Riders,Venkatesh Iyer,23.75,Rinku Singh,13.00
4,Lucknow Super Giants,Rishabh Pant,27.00,Nicholas Pooran (West Indies),21.00
5,Mumbai Indians,Jasprit Bumrah,18.00,Suryakumar Yadav,16.35
6,Punjab Kings,Shreyas Iyer,26.75,Yuzvendra Chahal,18.00
7,Rajasthan Royals,Yashasvi Jaiswal,18.00,Sanju Samson,18.00
8,Royal Challengers Bengaluru,Virat Kohli,21.00,Josh Hazlewood (Australia),12.50
9,Sunrisers Hyderabad,Heinrich Klaasen (South Africa),23.00,Pat Cummins (Australia),18.00


--- 
# Q6: Calculate the percentage contribution of each player's price to the team's total spending

We can first find the ratio by dividing the player's price_in_cr to the team's spending then calculate the percentage. We can use window function to find team's spending.

In [48]:
%%sql
SELECT * FROM ipl_players LIMIT 5;

,player,price_in_cr,type,acquisition,role,team
0,Ruturaj Gaikwad,18.00,Indian (capped),Retained,Batter,Chennai Super Kings
1,Ravindra Jadeja,18.00,Indian (capped),Retained,All-rounder,Chennai Super Kings
2,Matheesha Pathirana (Sri Lanka),13.00,Overseas(capped),Retained,Bowler,Chennai Super Kings
3,Shivam Dube,12.00,Indian (capped),Retained,Batter,Chennai Super Kings
4,MS Dhoni,4.00,Indian (uncapped),Retained,Batter/Wicketkeeper,Chennai Super Kings


In [54]:
%%sql
SELECT
    team
    , player
    , price_in_cr
    , SUM(price_in_cr) OVER(PARTITION BY team) AS team_spending
    , ROUND((price_in_cr / SUM(price_in_cr) OVER(PARTITION BY team)) * 100, 2) AS pct_contribution
FROM ipl_players
ORDER BY pct_contribution DESC;

,team,player,price_in_cr,team_spending,pct_contribution
0,Lucknow Super Giants,Rishabh Pant,27.00,119.90,22.52
1,Punjab Kings,Shreyas Iyer,26.75,119.65,22.36
2,Kolkata Knight Riders,Venkatesh Iyer,23.75,107.95,22.00
3,Sunrisers Hyderabad,Heinrich Klaasen (South Africa),23.00,119.80,19.20
4,Royal Challengers Bengaluru,Virat Kohli,21.00,119.25,17.61
...,...,...,...,...,...
223,Chennai Super Kings,Mukesh Choudhary,0.30,119.20,0.25
224,Chennai Super Kings,Shaik Rasheed,0.30,119.20,0.25
225,Sunrisers Hyderabad,Atharva Taide,0.30,119.80,0.25
226,Sunrisers Hyderabad,Aniket Verma,0.30,119.80,0.25


**Answer:** It is facinating to know that some "super star" players can take up to 1/5 of the teams total spending!

---
# Q7: Classify players as 'High', 'Medium', or 'Low' priced based on the following rules:
- High: Price > 15 crone
- Medium: Price between 5 crone and 15 crone
- Low: Price < 5 corne

And find out for each team the number of players in each bracket

In [61]:
%%sql
WITH player_bracket AS 
    (SELECT
        *
        , CASE 
            WHEN price_in_cr > 15 THEN 'High'
            WHEN price_in_cr <= 15 AND price_in_cr >= 5 THEN 'Medium'
            ELSE 'Low'
            END AS price_brackets
    FROM ipl_players)

SELECT
    team
    , price_brackets
    , COUNT(*) AS number_of_players
FROM player_bracket
GROUP BY 1, 2
ORDER BY 1, 2;

,team,price_brackets,number_of_players
0,Chennai Super Kings,High,2
1,Chennai Super Kings,Low,18
2,Chennai Super Kings,Medium,5
3,Delhi Capitals,High,1
4,Delhi Capitals,Low,14
5,Delhi Capitals,Medium,8
6,Gujarat Titans,High,3
7,Gujarat Titans,Low,18
8,Gujarat Titans,Medium,4
9,Kolkata Knight Riders,High,1


---
# Q8: Find the average price of Indian players and compare it with overseas players using a subquery:

We can calculate average price group by type but within the type we can see there's capped and uncapped, in addtion to this, there are also some inconsistency in the class. so we first have to combine them into 2 classes only (Indian and Overseas).

In [63]:
%%sql
SELECT
    DISTINCT type
FROM ipl_players

,type
0,Overseas(capped)
1,Indian (capped)
2,Indian (uncapped)
3,India (uncapped)
4,Overseas (uncapped)
5,India (capped)
6,Overseas (capped)


In [75]:
%%sql
SELECT
    player_type
    , ROUND(AVG(price_in_cr), 2) AS mean_price
FROM 
    (SELECT
            CASE WHEN LOWER(type) LIKE 'i%' THEN 'Indian' ELSE 'Overseas' END AS player_type
            , price_in_cr
        FROM ipl_players) AS subquery
GROUP BY 1;

,player_type,mean_price
0,Overseas,5.64
1,Indian,4.97


On average, oversea players's price is 0.67 crone higher than that of Indian players.

---
# Q9: Identify players who earn more than the average price of their team:

For this question we will find the average price for each team then filter out players whose price is higher than team's average.

In [76]:
%%sql
SELECT *
FROM ipl_players LIMIT 5

,player,price_in_cr,type,acquisition,role,team
0,Ruturaj Gaikwad,18.00,Indian (capped),Retained,Batter,Chennai Super Kings
1,Ravindra Jadeja,18.00,Indian (capped),Retained,All-rounder,Chennai Super Kings
2,Matheesha Pathirana (Sri Lanka),13.00,Overseas(capped),Retained,Bowler,Chennai Super Kings
3,Shivam Dube,12.00,Indian (capped),Retained,Batter,Chennai Super Kings
4,MS Dhoni,4.00,Indian (uncapped),Retained,Batter/Wicketkeeper,Chennai Super Kings


In [78]:
%%sql
SELECT
    *
FROM
    (SELECT
        team
        , player
        , price_in_cr
        , AVG(price_in_cr) OVER(PARTITION BY team) AS team_average_price
    FROM ipl_players)
WHERE price_in_cr > team_average_price;

,team,player,price_in_cr,team_average_price
0,Chennai Super Kings,Ruturaj Gaikwad,18.00,4.7680000000000000
1,Chennai Super Kings,Ravindra Jadeja,18.00,4.7680000000000000
2,Chennai Super Kings,Matheesha Pathirana (Sri Lanka),13.00,4.7680000000000000
3,Chennai Super Kings,Shivam Dube,12.00,4.7680000000000000
4,Chennai Super Kings,Devon Conway (New Zealand),6.25,4.7680000000000000
...,...,...,...,...
71,Sunrisers Hyderabad,Travis Head (Australia),14.00,5.9900000000000000
72,Sunrisers Hyderabad,Nitish Kumar Reddy,6.00,5.9900000000000000
73,Sunrisers Hyderabad,Mohammad Shami,10.00,5.9900000000000000
74,Sunrisers Hyderabad,Harshal Patel,8.00,5.9900000000000000


---
# Q10: For each role, find the most expensive player and their price using a correlated sub-query

A correlated sub-query is a sub-query that references column(s) from the outer query. We will get the player information and apply the filter in the WHERE clause to match only players with the highest price for each role. To do this we will use the subquery in the WHERE clause - the subquery will return the MAX(price_in_cr) with the matching roles from the outer query.

In [83]:
%%sql
SELECT
    role
    , player
    , price_in_cr
FROM ipl_players o
WHERE price_in_cr = (
    SELECT MAX(price_in_cr)
    FROM ipl_players
    WHERE role = o.role
)
ORDER BY role;

,role,player,price_in_cr
0,All-rounder,Ravindra Jadeja,18.00
1,Batter,Venkatesh Iyer,23.75
2,Batter/Wicketkeeper,Rishabh Pant,27.00
3,Bowler,Arshdeep Singh,18.00
4,Bowler,Yuzvendra Chahal,18.00
5,Bowler,Pat Cummins (Australia),18.00
6,Bowler,Rashid Khan (Afghanistan),18.00
7,Bowler,Jasprit Bumrah,18.00
